### ISSUE TITLE
🐛 Bug Report: Bitbucket Cloud integration maxes out at 900 repositories

### ISSUE THREAD

**Issue by benjidotsh** [backstage/backstage#26854]
### 📜 Description

We have been struggling with a fluctuating amount of components for a while now, with some components being suddenly unavailable, and we have narrowed it down to an issue with the Bitbucket Cloud integration.

As far as I can tell, this happens because Bitbucket Cloud's API ([specifically this one](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-other-operations/#api-workspaces-workspace-search-code-get)) doesn't go past 1000 repositories.

In page 9, the property `next` is missing. It's possible to go to page 10 by calling the endpoint manually, but as soon as you try to go page 11 it returns "Cannot fetch results past result number 1000".

Same goes for pages with a higher `pagelen`: for some reason the maximum `pagelen` is 373, so if we use a `pagelen` of 373, it works until we try to go to page 3, since it exceeds 1000 results.

I'm not sure how to resolve this since it sounds like a limitation in Bitbucket Cloud's API, so I would love to discuss possible solutions in this issue. As soon as we come to a conclusion, I'd be more than happy to submit a fix.

### 👍 Expected behavior

It should retrieve all repositories

### 👎 Actual Behavior with Screenshots

It actually maxes out at 900 repositories

### 👟 Reproduction steps

- Use [Bitbucket Cloud Locations](https://backstage.io/docs/integrations/bitbucketCloud/locations)
- Have more than 900 repositories in Bitbucket

### 📃 Provide the context for the Bug.

`{"class":"BitbucketCloudEntityProvider","level":"info","message":"Committed 900 Locations for catalog files in Bitbucket Cloud repositories","plugin":"catalog","service":"backstage","target":"bitbucketCloud-provider:default","taskId":"bitbucketCloud-provider:default:refresh","taskInstanceId":"41b49bcf-799c-4697-9980-d843c2419471"}`

### 🖥️ Your Environment

```
OS:   Darwin 24.0.0 - darwin/arm64
node: v20.15.1
yarn: 4.5.0
cli:  0.27.1 (installed)
backstage:  1.31.1

Dependencies:
  @backstage/app-defaults                                          1.5.11
  @backstage/backend-app-api                                       0.7.9, 0.9.3, 1.0.0
  @backstage/backend-common                                        0.17.0, 0.21.7, 0.23.3, 0.24.1, 0.25.0
  @backstage/backend-defaults                                      0.4.4, 0.5.0
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.18
  @backstage/backend-plugin-api                                    0.6.21, 0.7.0, 0.8.1, 1.0.0
  @backstage/backend-tasks                                         0.5.27, 0.6.1
  @backstage/backend-test-utils                                    1.0.0
  @backstage/catalog-client                                        1.7.0
  @backstage/catalog-model                                         1.7.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.8
  @backstage/cli                                                   0.27.1
  @backstage/config-loader                                         1.9.1
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.15.0
  @backstage/core-compat-api                                       0.2.8, 0.3.0
  @backstage/core-components                                       0.12.5, 0.13.10, 0.14.10, 0.15.0
  @backstage/core-plugin-api                                       1.9.4
  @backstage/dev-utils                                             1.1.0
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.9
  @backstage/frontend-plugin-api                                   0.6.7, 0.7.0, 0.8.0
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.31
  @backstage/integration                                           1.15.0
  @backstage/plugin-api-docs                                       0.11.9
  @backstage/plugin-app-backend                                    0.3.74
  @backstage/plugin-app-node                                       0.1.25
  @backstage/plugin-auth-backend-module-atlassian-provider         0.3.0
  @backstage/plugin-auth-backend-module-auth0-provider             0.1.0
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.2.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-server-provider  0.1.0
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.3.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.3.0
  @backstage/plugin-auth-backend-module-github-provider            0.2.0
  @backstage/plugin-auth-backend-module-gitlab-provider            0.2.0
  @backstage/plugin-auth-backend-module-google-provider            0.2.0
  @backstage/plugin-auth-backend-module-guest-provider             0.2.0
  @backstage/plugin-auth-backend-module-microsoft-provider         0.2.0
  @backstage/plugin-auth-backend-module-oauth2-provider            0.3.0
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.2.0
  @backstage/plugin-auth-backend-module-oidc-provider              0.3.0
  @backstage/plugin-auth-backend-module-okta-provider              0.1.0
  @backstage/plugin-auth-backend-module-onelogin-provider          0.2.0
  @backstage/plugin-auth-backend                                   0.23.0
  @backstage/plugin-auth-node                                      0.4.17, 0.5.2
  @backstage/plugin-auth-react                                     0.1.6
  @backstage/plugin-bitbucket-cloud-common                         0.0.0-use.local
  @backstage/plugin-catalog-backend-module-aws                     0.4.2
  @backstage/plugin-catalog-backend-module-bitbucket-cloud         0.3.2
  @backstage/plugin-catalog-backend-module-github                  0.7.3
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.2.0
  @backstage/plugin-catalog-backend-module-unprocessed             0.5.0
  @backstage/plugin-catalog-backend                                1.26.0
  @backstage/plugin-catalog-common                                 1.1.0
  @backstage/plugin-catalog-graph                                  0.4.9
  @backstage/plugin-catalog-import                                 0.12.3
  @backstage/plugin-catalog-node                                   1.13.0
  @backstage/plugin-catalog-react                                  1.13.0
  @backstage/plugin-catalog-unprocessed-entities-common            0.0.4
  @backstage/plugin-catalog-unprocessed-entities                   0.2.8
  @backstage/plugin-catalog                                        1.23.0
  @backstage/plugin-devtools-backend                               0.4.0
  @backstage/plugin-devtools-common                                0.1.12
  @backstage/plugin-devtools                                       0.1.18
  @backstage/plugin-events-node                                    0.3.10, 0.4.0
  @backstage/plugin-home-react                                     0.1.17
  @backstage/plugin-home                                           0.7.10
  @backstage/plugin-kubernetes-common                              0.8.3
  @backstage/plugin-org                                            0.6.29
  @backstage/plugin-permission-backend                             0.5.49
  @backstage/plugin-permission-common                              0.7.14, 0.8.1
  @backstage/plugin-permission-node                                0.7.32, 0.8.3
  @backstage/plugin-permission-react                               0.4.26
  @backstage/plugin-proxy-backend                                  0.5.6
  @backstage/plugin-scaffolder-backend-module-azure                0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.3.0
  @backstage/plugin-scaffolder-backend-module-cookiecutter         0.3.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.2.0
  @backstage/plugin-scaffolder-backend-module-gitea                0.2.0
  @backstage/plugin-scaffolder-backend-module-github               0.5.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.5.0
  @backstage/plugin-scaffolder-backend                             1.25.0
  @backstage/plugin-scaffolder-common                              1.5.6
  @backstage/plugin-scaffolder-node                                0.4.11
  @backstage/plugin-scaffolder-react                               1.12.0
  @backstage/plugin-scaffolder                                     1.25.0
  @backstage/plugin-search-backend-module-catalog                  0.2.2
  @backstage/plugin-search-backend-module-elasticsearch            1.5.6
  @backstage/plugin-search-backend-module-explore                  0.2.2
  @backstage/plugin-search-backend-module-techdocs                 0.2.2
  @backstage/plugin-search-backend-node                            1.3.2
  @backstage/plugin-search-backend                                 1.5.17
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.8.0
  @backstage/plugin-search                                         1.4.16
  @backstage/plugin-signals-react                                  0.0.5
  @backstage/plugin-techdocs-backend                               1.10.13
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.14
  @backstage/plugin-techdocs-node                                  1.12.11
  @backstage/plugin-techdocs-react                                 1.2.8
  @backstage/plugin-techdocs                                       1.10.9
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.12
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.6.0
  @backstage/theme                                                 0.2.19, 0.5.7
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.9
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by benjidotsh** [comment]
In the meantime, I have found a solution that works for us by first retrieving all projects and then calling the endpoint for each project. This moves the limit from 900 repositories per workspace to 900 repositories per project.

I'm thinking of wrapping up a PR that allows you to choose some sort of mode to allow this.

**Comment by benjdlambert** [comment]
Oh wow, well that's not great is it? Seems a little strange that you can only get half the results back with BitBucket API's, there must be a better way to query the API to get a list of all them for sure. See what you can come up with and let's take a look in a review :pray:

**Comment by benjidotsh** [comment]
@benjdlambert if I'm understanding the code correctly, we're specifically searching for repositories that contain a `catalog-info.yaml` file, which also helps with creating the location of this file, so I'm not sure if there is a good alternative.

My PR should be ready today and then I'm off for 2 weeks, just a heads-up in case I don't respond 😅

**Comment by pjungermann** [comment]
Unfortunately, there is an issue with the Bitbucket Cloud API and its pagination that I tried to address with Atlassian. So far, they were not very helpful on that one.

The issue is that the result set has no stable order and the pagination is page-based. Hence, going through all pages does not mean you will get all the results. You might have received some results more than once and others not at all.

I was already thinking about a solution for that one...

**Comment by benjidotsh** [comment]
This specific issue has been fixed in #26873, so I guess it can be closed? @pjungermann @benjdlambert

**Comment by Rugvip** [comment]
Gonna close as fixed by #26873, let's handle other potential bugs in new issues
